src/app.ts

```diff
@@ -0,0 +1,366 @@
+     */
+    public async getUpgradeableContracts(): Promise<
+        {
+            address: Address
+            wasUpgraded: boolean
+            implementation: Address
+        }[]
+    > {
+        // Sometimes the accessList includes the rateProvider, sometimes it does not.
+        // We need to ensure that the rateProvider is included in the list
+        const allContractAddresses = this.accessList.accessList.map((a) => a.address)
+
+        if (!allContractAddresses.includes(this.rateProvider)) {
+            allContractAddresses.push(this.rateProvider)
+        }
+
+        const proxiesWithRateProvider = await this.getProxyInfo(allContractAddresses)
+        const filteredProxiesList = proxiesWithRateProvider.filter((p) => p.Proxy === '1')
+        const proxiesWithRateProviderDeploymentInfo = await this.getDeploymentBlocks(
+            filteredProxiesList.map((p) => p.address),
+        )
+
+        const publicClient = createPublicClient({
+            chain: this.chain,
+            transport: http(),
+        })
+
+        const mergedInfo = proxiesWithRateProviderDeploymentInfo.map((d) => {
+            const proxyInfo = proxiesWithRateProvider.find((p) => p.address === d.address)
+            return { ...d, ...proxyInfo }
+        })
+
+        const receipts = await Promise.all(
+            mergedInfo.map(async (contract) => {
+                const receipt = await publicClient.getTransactionReceipt({
+                    hash: contract.deploymentTxHash,
+                })
+                return { ...contract, receipt }
+            }),
+        )
+
+        const proxiesWithRateProviderCompleteInfo = await Promise.all(
+            receipts.map(async (info) => {
+                if (!info.ABI) {
+                    throw new Error(`ABI not provided for contract at address ${info.address}`)
+                }
+                const events = parseEventLogs({
+                    logs: info.receipt.logs,
+                    abi: JSON.parse(info.ABI),
+                    eventName: 'Upgraded',
+                })
+                const wasUpgraded = events.length > 0
+
+                return { ...info, events, wasUpgraded }
+            }),
+        )
+
+        return proxiesWithRateProviderCompleteInfo.map((r) => {
+            if (!r.Implementation) {
+                throw new Error(`Implementation is undefined for contract at address ${r.address}`)
+            }
+            return {
+                address: r.address,
+                wasUpgraded: r.wasUpgraded,
+                implementation: r.Implementation,
+            }
+        })
+    }
+
+    /**
+     * Checks if the rate provider is upgradeable.
+     * @returns True if the rate provider is upgradeable, false otherwise.
+     */
+    public async getIsRateProviderUpgradeable(): Promise<boolean> {
+        const upgradeableContracts = await this.getUpgradeableContracts()
+        return upgradeableContracts.some((c) => c.wasUpgraded)
```

Comment:
Is this indeed a good way to detect if it's upgradeable? I mean, is it possible that it's upgradeable, but it wasn't upgraded?

Reply:
This needs to be reworked.